### ISSUE TITLE
Large enum causes React hook deps error TS2590 Expression produces a union type that is too complex to represent

### ISSUE THREAD

**Issue by brieb** [microsoft/TypeScript#41615]
We use React hooks with TypeScript and ran into an issue where we were passing a large enum into the dependency array of the hook function, resulting in `TS2590 Expression produces a union type that is too complex to represent`

This is an example type signature of a hook
`function React.useEffect(effect: React.EffectCallback, deps?: React.DependencyList | undefined): void`

Even though the deps type is `React.DependencyList` which is `type DependencyList = ReadonlyArray<any>;` it appears to be attempting a union `deps: (MyEnum1 | MyEnum2 | MyInterface)[]` even though the type is ultimately discarded.

This gives the same error `const deps: any[] = [MyEnum1.Key0, MyEnum2.Key0, obj];`

We could do `[MyEnum1.Key0 as any, MyEnum2.Key0 as any, obj]` but that is not ideal and also breaks the rules of hooks eslint rule.

Another workaround we came up with was `[MyEnum1.Key0, MyEnum2.Key0, obj] as const` to avoid the unioning (though this fails the hooks eslint rule which sounds like a separate issue).

Thanks in advance for your help.

**TypeScript Version:**  4.0.5

**Search Terms:** react hooks deps, dependency array, union type that is too complex

**Code**

https://gist.github.com/brieb/ef53a9e48e8a5dec6cfa7d2cfda22623

**Expected behavior:**

No compiler error

**Actual behavior:**

TS2590 error: Expression produces a union type that is too complex to represent.

**Playground Link:** 

[Link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AocpAOwFc8BZATwFE6QBGOAb3LgIDSSJgAZ8cALxwRAGj6DhHcVI5z++IUwBMyuFrUKmAZl1GDG4QBZdl85oCsu+3eEA2Xa5dMA7Lu9eADl0ArwBOXVCvDjFJOGiopViOVXkLJg4dJP1UzQ4TJLMcxWsk2yL0xyTnco53JM8a3yT-GqCkkJrwpMjyrRipPq8tRIGU9U0tTIHs8eEtfIHC2e0SgbLlrUqB6o26gYaNpoGWjbaBjo2ugZ7lo364O68jEYexwyMph5n3hYel99WD3W7y2Dx27z2DwO7yODxO7zODwu7yuDxuhks90xXksL1xOM+lm+aUsv1JOMBlmBJNBlnBJMhlmhJNhlnhJMRlmRJNRlnRaXs90FXnsL1FIs+9mJDl+9n+AsB9mpDlB9npDkh9mZDlh9nZDkR9m5DlR9n5mlc90tXlcL1tNs+rmlbl+rnlFsBrmVblBrnVbkhrm1blhrn1bkRrmNblRrnNwm890TXm8L1TKc+3mdPl+3ndCcB3m9PlB3n9Pkh3mDPlh3nDPkR3mjPlR3njTAC907gReATeaQCn0HgV+AXzHcBAWLAVBM8CkIC1YCsOXgURAWbAVRW7C91CsnKoReR7Cn1C2dCv0vYUBoWLoVBD7CkNC1dCsPfYURoWboVRf6iER7miA9lmiF5wMAz5omzaJfjgwDAWiYtolBVDAMhaJq2iWEcMAxFombaJUWIhJgPiGpkl0ZJ+1yDJqIyBJ4LyBIkI4FCODQziEkw2oElwjh62SAiOCIjgSPEqI+mowYamGGTaMUSYZNg+YZPHDIkK0FDNhk8sMkwrRsK0XCTKkgitCIrQSOsqI7mox4ameBzFPSD4HNgox4K8uykKMFCjDQwK7MwoxsKMXCIrsgijCIowSPiqJMWo7EalxFLXPY6CiSS+DyTSpCqSStC6SSzCmSS3C2SSgiuSSki+SiQVqOFGpRRazKpRa2C5RajSlRalC1Ra-StRa7C9RaoSjRaoizRa9tamA60altaj7VW6CnSiN11o0r11pQv11v0oN1uwsN1qEqN1qIuN1sWxNqOTRoIPTRpoKzKI82ejSi2elCy2e-Sq2e7C62eoSm2eoi22exbO2o7tWggvsokHJHYLHJGNKnJGUJnJH9MXJHsOXJGhI3JGiK3JHFv3aj9yiI9Gcy89Gdgy9GY0u9GZQh9Gf019Gew99GaEn9GaIv9GfbPp7nloYRBePpXL6T51aV34+nHPpAT1pXQT6cs+khU2ldhPp6z6REbaV1E+jl6JdGGUDDGGFXkiGejYmGbNhm1ljenYl32O9o3uODs2+ODy3BO923RO9h3JN6aTfdkjZ5IztXlIz-21Iz3XtJd7Shl0jOTaM0vq0mS2zLT23LPLh3bN6ezfccjZnM7tX3M7-2vJdnz2-1-yhkC4eTbC4fa4i4frdi4fm3mB3Et6ZLfdSjZ0q3tWiRdnKN+1-Kd-1oqN6N0qN7NiqN8t6qN9tuqN4dxrema33Wo2dqv7Vrqv7+16l-XWA0v7Fk2EbNUQwxpf1rpNL+1sZpfxXvNL+ctLQuxWrsFWG1dga22r0Xavs3RDAOiQiBx0SEmzOiQ2ul0SHWxuiQle90SFyyer7F6hwVbvUOBrL6vQfpcN1v9LhECgZcJNqDLhtcIZcOttDLhK84ZcLlojX2yNTgqzRr0DGmj-bY00brPGmiIGE00SbEmmja7k00dbKmmiV6000XLBmvsma9BZh4tW7MPH+y5h43WvMPEQIFh4k2wsPG1zFh462ksPErxlh49sdx7hpKeMrUwytMmfDuNmO4vxCmZMBHcYsdxQQVMyZCO41Y7iwnqZkxEdxmx3FRG0p4ztYjPDdmkZ4Lx+mdLyYxcozwilB1uCHbpYdRmcVMJHSZNSY6TIafHWZzSk6zPaanW46cpAfF6ZoD4AzhhPDzvsyYZyinzDOaUsuoyK4XPLB8GpRkzkNIbrs5pzcHntLbrcDu+yu4-AGc8J4-cgUFKHt0keALSnj1GZPGFzyZ4wrqfPGF9YvLNNiuC9p69bib32dvAEAz8SjIPt0o+hKimnwBKUi+hLKnX0JTUu+hKGmP0Jc0l+hL2nv1uJ-fZ38QQDPFKMgBwqCnAOFeOQKpSlRPGGt06BEqalaiVQ0vUSrmlGiVe0s0TxMHdOwRCAZeCIR5MIbcYh+zSGjPIXa8pVC7XPNoXaupDC7VYuYXa1pbC7WpM4fs7hMIBl8JhHkwRtxhEhrlWIkN5TJEhueTIkNdT5EhqxUokNrTVEhtSRo-ZWiEQDN0bcfRxaClGOLXK0xxbykWOLc86xxa6l2OLVixxxbWkuOLak9x+zPG3G8UO1y8U8nnieIEodcqQlDvKeEodzyolDrqbEodWKElDtackod7ZMRYiAjiLJsRMSuUxISEQ2ZMRkhEOOTElIRDFkxLSEQ5ZMSMhENWTErIRD1kxJyEQzZMS8hEAerpUhcSHKsFRM9Xtyi4kJCM5YuIyQTIxFMqDMzUNzPgx+2oNhlmYdZGs3DnJNm4d5DsjEey4BEhg0wIkeJTmIfOfRy5bGyQ3LY5Se5qHHkcY-dXM9by2Osk+bRzkPyBO8n+RiQF9HgUkh7lBsFiGIVKZvdCtTD7-I2ARahpFamP2orUz+jFamANLzPbijTvICUYiJfRklJJd5QfJahylHmb2khsHStzlJGVOdpCypzjJ2VOdZFypznJeVOd5AKjEQr6MippHicVqHJWpZvTK1LD6wFQcVYh5VRWP2wKKz+hBRWAPIKKyBtBRWD3Gqg6ahkeILUMkJNajEtr6P2tQ46-rL6XX9Y-e6-rP6vX9YA76-rIGA39YPcG+joaWR4gjSyQk0aMSxtWw+hNq2X3JtWx+tNq2f2ZtWwBnNq2QP5tWweot9GS0cjxOWjElaXs3prS9h99aXsvqbS9j9raXs-o7S9gD3aXsgb7S9g9g76PDoxKO5HF6-FQanYhmdyOH3zuRy+pdyOP2ruRz+jdyOAPbuRyBvdyP2yCiFMe8ogoxQ5NZ9epw16RT3u5+OQUipn289VO+3nmpv2891P+3nhpgO89NOBkUkG4CikY6KMUCHliiklChwwopZQYYFFh1XOH9d4akKKcsopNTEeN7qMj5vDSUfN6aGjAo6NSnV9nS3rHtfsalNmKUsoeP+8VPx-XgmpTW5E776sUpdSSY94aGTkfTTyYFIpuU6vVOq-U9rzTcog86bzwL-TsQ5TFjlKqYKrOzN5-j5ZvP9Y5SGjswX00jmFRClcw4dzqvPP6+8wPoPfmK8Bb74qYLCpVRhYVJqSLCpdQxYVIaeLCpTRJYFCltU6vf6W8y-r7Laog95bVALwrqviva9K1f63FWr-x+q1flvdWr-NjVKaQ1rOWuq7axqMUTrDUSUHrAUPrLUAXIbLUKvUbLUa3CbLUePabLUFvObLUD-RbLURnFbPUdXVMJwTbHUSUHbAUPbPUAXQ7PUKvE7PUa3c7PUePK7PUFvW7PUD-B7PURnZ7I0dXPsJwD7AUL7I0IPX7I0AXAHI0KvYHI0a3MHI0ePSHI0FvGHI0D-eHI0RnJHM0dXNHM0VyM0SUbHbXXHM0AXAnM0KvYnM0a3MnM0ePSnM0FvGnM0D-enM0dsS0K0FnZYS0O0DnHwrnWIS0bMS0V0e9G0Z9DwYXcoS0X0MXGI79KI6sS0UMGXBIyMeXBI2MJXGIlXW0RjW0O0LXQwW0R0PXNIW0V0I3C0E3W0YsW0X0BZEowjIIu3Go0MR3Co0SDwF3Lo2Md3C0OjJ0Aon3OAJ0VyJ0R0TjHwwuKQJ0ccJ0T0CPCowTJ0csJ0QMMTGY0MJPQYyMVPFY2MDPC0RTN0Ao3PN0CYzTN0EIkvN0BY8vOYwzEo4zMY2vHw+vN0ZIpvN0esN0SMdvF42MLvD0K0XvNwfvL0CY4fL0EIsfOYifCEz0afD0X0OfD0QMRfD0UMFfD0SMdfD0WMLfC0FLP0Ao-fMYw-Co7LP0EIvLP0BYy-P0eo2-P0dYh-P0ZI5-P0P4t-P0ZsP0WMb-Hw3-IMAotaIIoAgMR0UAi0PrIMBYobIMeo0bIMdYibIMZI6bIMP4ubIMAUxbIMDwlbMMAo-AoIwgkMR0Egi0PbMMBYw7MMeok7MMdY87MMZIq7MMP427MMAUh7MMDw57KMAovgoIgQi0L7KMEI37KMBYgHKMeo4HKMdYsHKMZIyHKMP4mHKMAU+HKMDwpHOMAotHOMCYzHMYowko3HOMBYgnOMeo4nOMdYsnOMZIynOMP4mnOMAU+nOMdsRMJMbwwwRMNMfwkcwIqQRMbMRMXMcI8oRMQsaI5YRMUseIlcxI2IRMasRMWsNIjcxsTIjc1sHIlclXVMRjVMNMYotIVMTMcozQVMXMaohME3VMYsVMUsJo28loqcto182sTox87orc3o4C1sAYhMOjLMS80YrMVyLMTMaYkc2YuALMccLMQsZYx8wTLMcsLMSsLYlC2sXYqCxsA4nC1sY4hMRTPMS83PPMBCzTPMGckvPMDCx4tC542814vMfC+vPMHcpvPMesPMRsQEni1sEEgsJMcEnwfvIsBC4fIsGc+EtCxE+SwsFEgsUsdEgsSsLEgsWsXEgsRsAkgsVsYkhMFLMsS8ikssBC7LMsGcvLMsDCy-MsD82-MsfCh-MsHc5-MsUSt-MsZsMsVsYUkc3-KsS8iUqcqUisTMWUhMPrKsDCobKsD80bKsfCibKsHc6bKsUSubKsMKxbKsAclbOsS880qcy0msTMG0hMPbOsDCw7OsD8k7OsfC87OsHcq7OsUS27OsMKh7OsAc57JsS8sMqciMhML7JsGc37JsDCgHJsD84HJsfCsHJsHcyHJsUSmHJsMK+HJsAcpHNsS8tHNsBCistsGc3HNsDCgnNsD84nNsfCsnNsHcynNsUSmnNsMK+nNsdsTsLsYcgcU9KQTsVyTsIcHncoTsUcec5YTsScZcwwTsWcdczGzc6GyXRG-9YIfc3G9cI83G7cU8zGlXPsRjPsXsG8zQPsIcB84QPsUcF8jsE3PsacC3OAPscsPsBcf8rmlcICtmkC6GsCiW7cSCjsOjQcOm0YwcWG9jQcbMQcUcUPTGkuWIQcacQTQcQW2PfmoigcEyYIUi+W9cCitm6yS20GxTMcOm3PMcWGzTMcDWkvMcccMcScbipm14scQW+vMcJcJvMcesMcdcCSwO7caSicLsOSqcXsQfAcYfKcDWtSqcX2qkYIbSicWcPSicBcQyicFcEyicdccyicbcKyjsFLGcOmikmcWG7LGcDWvLGcX2y-GcacW-GcQWh-GcJcZ-GcKOt-GcTcRrfmyKgcX-RcOmuK-mhKxcIcZKjsPrRcX2obRcacUbRcQWibRcJcabRcKOubRcTcRbRcUGlbZcOm2q-m+q5cIcJqjsPbZcX2w7ZcacE7ZcQW87ZcJcK7ZcKO27ZcTcB7ZcUG57DcOmma-muajsL7DcDW37DcX2gHDcacYHDcQWsHDcJcSHDcKOmHDcTceHDcUGpHLcOmtHLcWGisrcDW3HLcX2gnLcacYnLcQWsnLcJcSnLcKOmnLcTcenLcdsfcPcCGzQfcY8cctIfcM8BG5YfcK8FGwwfcW8DGpR99CIHGvRl8AmtRom2Ifcesfcb8cmvR-8KmpRlXI8RjI8Y8Rm4QI8M8VmpgI8K8Tmo8W8M3Bxx8b8uR38uAI8N8QSCIcW7xyW8J6W2J-8OW88PcTOLR0Y88Vyc8M8ZCpR1C88ccc8W8bC9xwTc8csc8F8M2uRi28xq288b8W27x+2upqRxTS8Zx3PS8LJzTS8C8EvS8Qpziy8e8V4y8Cp+vS8N8JvS8SxmzKQS8X8eKCIBOu8PcOSu8Y8NOuR4fO8C8NSu8QpvO8xguu8R8Yuu8F8Muu8D8Suu8b8Guu8f8euh8PcNLORikh8LJ7LB8C8PLB8Qpy-B8e8W-B8Cph-B8N8Z-B8Sxt-B8X8Geh8KR3-V8Zx5e18LJp0CIDe18K8AbLRobV8e8UbV8CpibV8N8abV8SxubV8X8RbV8KRlbd8Zxp+98LJrMCId+98K8PMMIQ7d8e8E7d8Cp87d8N8K7d8Sx27d8X8B7d8KR57H8ZxxBn8LJr7H8C8X7H8QpgHH8e8YHH8CpsHH8N8SHH8SxmHH8X8eHH8KRpHP8ZxtHP8LJisv8C8XHP8QpgnP8e8YnP8CpsnP8N8SnP8SxmnP8X8enP8HoAAX0oBoHoDgGYDYHoC0B4HKHuEYwgiGBdieFMBxBsBFCcBtA8BTD8ECGCDCAiEAmomzYgkymgk8gciShSiahah2nWm+menRiRmZkZiVhdkY1GL93dg1kHmHiGHPhgRdhoSwSGEtgcRdiSRdlSXSUY1z3z3eDyShVMDlQZSVVMDdVMAzVMC7VMF3VMAPSPRxDxAvUJG0wMwpH8xxFpHGxsEuxsGhxsDpxsEZ2Z2VycH0MlGLycAFynxFFVHgKcEYIIJFENFcKcA8K8JtCKIdA8FuI8AWORJtF9HVI8E9I8GzI8F7I8AHKHJTGvIzD8FYr8Awq0pTFLFyr8H6r8EOr8CBr8FBvBp7GCFVstpHGCFzvzsCFnCPuCBAeCDIeCHEeCCkZkbCFcdPAiH6ZWZvAiBBYiDJexf5a5a-AiCjYiEWiAjrezahriEUdyEnIs9gj5ySE0bSGQjrZQn0fs-0jxos+wjMZUBJsc+AzraInAzrcWhV2SHrYbcyh9hUC8eSGYg0hN2SA4i4n0jCeSGwiiaSBieEgYjEgkkWjowyGzdGIyEbabdUngm1sc91pUBKfSEEwyH0hNoMiklMiEkshkishskWkUzyGzdzzyEyk0zyE8m8g0k4ryACiCn0nrzyHCkiiEnmbiFjsUGWYKEWmc3Ymzf73YkymHyylyhSg0mOZUALvYhKn0iZBSmwjZBSiEi5BSiIj5BSkWhS04mzYpM4k6mgilCangjlCaiQmv0ME4jQlVTAgf04gmlwm1TagIj1TahIjntyF-yWh2gggStqC2lgj61qH2iQi9B2jQj9B2kwiDB2lwjDB2gIijB2hIjjG+mAjW1yCfsEkyk5eaFgj20Ej+iQiLG+jQjLG+kwirG+lwjrG+gIibG+hIjbHRmAle1yEQdEkyi+1EixngjHHRiQinHRjQjnFaEwkXHRlwlXFaAIg3HRhIh3E6GAhR0c7R3EjZmgirPt-gmvE6CQjvGZjQifE6EwlfGZlwk-E6AIh-GZhIgAl6BM99lkbmHM9ViVg1lUfdjs4GAc4mEiJj4gVc7T5Ng8-Nij8tl84z9thsYz4dnsYmBV1dm9k9jVii70C8YDlDl1hN2GAgT5uGBNjCeGFrgy9GGtjieGBXnElDjljo0mGHZHdHbSEmHHYLm1kq4mGq70Fq8mCNk2HLjNmqbmFqemGtna4zhXhaemDlkU3mGHdz3mD7g1g+Anm1lhXdk4vmAgVePmGnjNjCgnktiinbltmW+0CrdFgcsZzNpGHb95tI+8DWNSndhqVtIusY7noALraQr4JsS7lvFrg3ct41se7lvBXhPct4csFLJsGHYUlNg-8DWD9w-jax-uH8fWED1n635IEMCM2Bqg-iWwYeP8W2PDx-gOwkecwX-EZGHbL0jIasLFiQn9h9YjIusIbEZEoRGxieRCM2GTyISWxKeRCW2DTyIQOx6eQiBWEzzmBP0TIasdnscH9h7YTIoifWHzyERGxBeQiM2CLyESLtFEtsKXkIgdiy89ECsBXnMEQaWQ1YX2SyIYm1ia89E+sHXnoiNj69TgZsI3noktim9TgtsC3nogdjW9LgCsO3hMDRzWRfEGsF3jkO1ju9Lg+sL3l4iNi+9LgZsAPl4ktjB9LgtsMPl4gdiR9bg0ffZLH2MDmc7g46azvkkyRFJ0+wgMpNknKQ58HghjI5B51qSZIGkxfEYf526Rl9Fh7SSviMJVw9JOkAyNxsYAb7PACkeQeZHKhNzPBykfNZ4M8jCbPA6k-fV4FijibPBWko-aZKkjowHIzkJycdOxg+AFJUKHwOVCvw+DlJBMHwZ5CbReTvJTAVtD4N8laQn8vgqSRTF5A3abst2fSTTF5ChRFJH+GI+FOUleJeQUUNSL-oigaS-8AUOKVpEAL+CpJnM-kDdv3n8jjph8-kApGpX8j7sGU5SOkKYGLr+Q2UdSLAcSixS4DiUrSAgcSlSQpZAoG7CkoFHHTZZAo0qIpLQMFQKpykt+QKM8gfyBQ6kz+QKFijfyBRWkM9QKKkl-xhQN2y9MKOOnEF2oCkfWMKHKiGxhRnUlSRQTahqQqCbUDSdQTamaRaCbU7SXQTGnSQGDjAT9CKOOlMFwgCke2CKPGlKQ2CY0lSewTGhqROCY0DScXqMluwRQ807SLwRWnSS+DjAiDWKOOi+yxRq0RSMIRWlKSRCK0lSGIQiBqTxCK0DSJIQiGaSpCK07SDISiHSTZCRhaOeKOOgrITpp0RSEoSiFKTlCR0lSKoSiBqS1CR0DSBoSiGaTNCR07SNoRiA6H0Yuh56GwJZysDWdL0J6O9A+kz5QZdGmgV9KeI-QedP0J6P9ABiWF3iQMQXM9OsKYwq5oMt7O9hegb5IZb26GB9CblxAvo+auIAjIyBFq4hSMAGOJriBAwvDsMB6OjAxgJAsZ72D7G9KhSJAPoV+RIF9IJiJDCZGQO-JjHvw4wAZD+UGJpkSDkwHpFMpIRjKSDxDoiHxmmUkNpjJC4jeJlIAOlYFeKkhTMjIUkUZlZAUiFMnIAAaSAcwHpnMVITif3ipAXph8VIXzGSA0pUggsL6XkWemLpUgIsP6YUS5gAxiiXMIGSUS5gPQpY6QnEiknSAvTZY6QuWMkGqOSyUgGBD42-HSHKyMg2BWWVkJwOSycgeByWXkPwKYy-4mQnE5ekyAvT2j+sN6PrEyAfRDYmQI2WkJ6N6yMgfRvWVkP6N6ycggxvWXkKGN2xYgIxbIDbBeljFsgb0e2NkAdkpApjdstIdMbtkZBZjdsrIXMahluxsh7svIYsZ9ixBliuQ72C9F9i5A-YyQ9Yz7JSCbGfZaQrYjkIyA7GfZWQ3YjkJyD7GfZeQg4nkFiBHFMY0cfIDHISEKFWBccfIfHJSAXGo5aQy4nkIyDXGo5WQm4nkJyB3Go5eQ+4gUIeKZy852c+hazoKCDyp9VcwwpgILm5xV4Jhgoa3B50FDx5vO8MlvJ+Phkf4fxluP8WriA7q44MluXYTriA5B4jhFeTmqKEVCBMHAfNK3Mrltzx47hooFvHE1FAf4MJpuRnJ7jSYe4xQM-BwAHlybiyQ8AuFflKCrxR5N+rOE2lKHjx0SE8EoFPB-gRFShGcWeZTDKDFA8ThAReMDsXllBCTjZnFSvCKFeLV4bZmoaSfribxygW8i3VvDbM7yM5nMSodXP3iVD6Fh8SoUfLKA0pKgp8VeYyZbmLpKgF88eSyUqBbw2SlQH+eyUqEZw753mxsikmqH0LH4qBN+WUD5O3yKh-JWcqDvfk1ChSj8uoCKdvkNDRTt8X+RnL-i1Dq5l6WofQmlK1BB5wC+LMAoqAJ4-5VQBUsApqGKlgFdQZUsAoaEqlgFTQNU0gkKAjF6gxQ9VPUMQSDxkE+WrOSgl1NIKqheppBTUANNIK6hhp+uVgu4O1wcFJpghIUGWKNBihkGwhfgiIVlArTBCiodaYIVVBbSDQmoXaYIV1AHSDQCHNQqaDOkmghQl0vQkh30IVkDCIoEwrOIFDmFXpqC1UB9JNCahvpqC3UH9JNCGhAZqC00CDItCHjPCERPwhMWs7BEIiYRBYreLGL3i3AEw2IhEUDAmMSi2MlIhEQyIClCZzCjwnkQog+FyZYxXYaUR6IhFaZcxTmnUR6L1E+aDRVDoGBFq2gOifxOJraAFL8zbQHhIYsLMGJ2gxZbgdjJMQdCugl+ZipYvUVWKKyZimxZInRKdB-FGJYxJpk6COIeFTi+sl0HaCNlMAbiWHW4q6AtlBLOKboeoq8TdDrFPijsioj8TkmJKASApGkW6A8LOYvQBRSEtswhKOgYBFRNSl6Gw7Il6ikcsYsXS9CYlkilkr0H8RslegBS9kr0B4VJKZyglFJP0BMRpL5ySi9JIuSSU9ClzOleHdkoGCrnUlQwtckkpGAbkkkhSHhUUv-gDB2gEqQYGUiEXlJ9y5SnoQeSKV9Ajy5SgYceXKVDBTy5SkYWeXKVjALzbSVoCMWGDtD1Uww1pEInaW3k+FHSe820r6EPm2lAwJ820qGHPkVFfSV8kogGVvmRkrQZYqMHaGQbRkSOMZV0B-MjKehv5kZX0H-IjCBhAFkZUMCAojCRhjpFRfMpApjBWhLpZZMjuWUdD3SglNZFBRaHrLoKWVvoLBTGEDC4KWVoYAhTGEjDEKWVsYMhQmEPGDkUw5nUcpKszDJ9bycM2cpKqXIfkJhq5SVZWC4Xyq9yolPGYmDCqCLEwA5c8qIpHLiKrylHe8jOVkVoVOa75PwEzNfJfl8KYTVMDuTuGphRKcTVMGFX5mpgBy0FIxVBTTCmKfA7GRChmFzDWLQ1WFD8rhQcUoVCKO5OiVmFEruKswYVBEVmAHK0U-FOYNMIEpYp0dWKuYcJXmELCiScwpYd4iOQEoJLHywlZJfWvEphUaReYAcs5iLCXkFKeS+SpmEKWPk1KRYejlpQ-IVKiw+FdAVOTLpFhjKolGyUWDCr2SiwA5Gyh0ocp+AqSj5Jyn0tvKuVBl1lQsCMp8obrfKlYSZVutrAzLrKjYeZdZQioDloqKyisGmASpVgkqM5VKjspSqFh9lUVUsEcpSqVhTlKVWsBcpSqNhrlKVVsHcuapJgIxdYNMPVTrCNUZyLVT5SOXao-LmqpYf5c1UrBArmqtYUFY+WGoQrbyY1aFfNSTBlimwaYZBotQ45LVcwaK+aoWExXzVSwOKhsJWHxXzVawRKhsI2FJWPlTqFKlsEmEuk3UuOt1TMAyqepcdnqhYNlQmHeqcqWwlYHlSptrD8qWwjYIVSptbCiqOwh4sGoEHM4w0zN8NDWnDKRpmb0a04CYVjTM0LhNVTNbGZ2Cjp4zOwm4QRZ2FBo00TVA4cRfTR7As0Na1q9mj2EnAOquas4EJhLWFpLg7hfYKOnEz7Cbh+ZfYUGgrUDXy1ewIa9WpbQ1qoVNagQFfvrTK2zh415tBcDRMHArgrag4G2puARGDhHaXYXNW7SE7u0hwd-RGt7XCV+0hO04IOtWoHCh061bNCOo2qm0x1NwNIscKDWcxTg6a-eFOoEAzr9q2a2dDSlOEnAF0pwRdQWpOv5pl0pwFdKOjZKnCbh7JU4UGo3Q6Wt1ggm6tmu3R3VM0u6+6hupOBGWD1ntQ9BcOete0rgr1DddcLeobrbg4pi9CTkvV7Cr00pa9ecKOC-Wb1Jwv6+erOAA2b0FwwGzeiuDA2b11wkGzetuBg0f0uwEY5cL2BfqxjX6gQT+uhoHA-0sNH9WcLho-oLgCNH9FcMRrZoQMyNTNaBpRpQZdgyxG4XsMgzQYyd0Go4FjSg0nDsaUGs4LjUQxk7EMVwAmihjJ0obbgxNdDOTvQ17AnhEazDBlewzk4cNJwymjsDwzU0CM5OgjFcDprEZydxG24IzdIwMbONzO8jXcCowvBwz1Gu4HRveAmH7gKmHnfcG+GxkWNdw1jX8IIv3BSNHGgW0Jq4yyYN8PGinXxoUxNz+NFOwTCpmEwiaKcPwMTI8N+ASZHgkmUjOjCk1PDHgQ12TFTheHyZRqimKne8GU2q01Mqmb4OieeEsbuKGmp4f8NRW8btNc13TFZj0zPB9a1GAzcJSMxWajNHwY2uRpM0m2T6PwM2yfd+AAGXh-wqzZzGsxvBbMsmuzLbd4wOYaU7wt4U5hUrOY3grmb4SyTcxvAPNfw9kp5mEBSyvM-9x4F7d4x+bvb3G-zL7cCy04gtHwYPLRhCyB0gGPwoO+FlpwRb-g4pqLbFmi2PAJVXwZ4HFn1lxbPhbwGOuRiS2x0UtsWlLD8ITrpbYt6W-4cnSyy5astjw9Vd8GeG5Z7YeW-LW8KzpFZctRWL4bnVKy5bStvwgu9xvKxF3KsDOKrY8Mg01YGctWV4eXfqwM4GtHwXG01gZzNYfgBN1rAzja3-BibHWRnJ1seBN1qM3WDKr1kZ29a3gbdAbIzoGxfCabvGobHTZGyM5Rt-wkfeNuQGADUAYASAKAJgHQBIBk2TAAAJIhGwjERtAFEd4D8AAAbigAAA2AALjgAABnGAFAGCMABzCgIEbQAQBqA+RuABAAABGAAKxyPMA4joR8I5EdiDcA4A6R7IwQAyD5BY2FAcgAAHpBjcAFgAAA8wARAXI7keAAVG4AkxiAAABNaASR3I3ABQBwBaA1AWY9QDgAwAmAYAKIzAAAAWKAeAMADWMwAIAEAOAOUfAAZGkAYxvYzcaICTGkAuRmgDAAAB0MAXIwAApNg+4AAJTkBkA6Ab47QA+MsBMAmAJABgD+N-GgTkgAAHw8BY2MgOAAAG0U27ADgF8c0CyBojqbEAFoHxPCBCTtRuowAF0gTFAIAA)

**Related Issues:** 
https://github.com/microsoft/TypeScript/issues/33130

**Comment by DanielRosenwasser** [comment]
> This gives the same error `const deps: any[] = [MyEnum1.Key0, MyEnum2.Key0, obj];`

It really sounds like the hooks thing is a separate issue and it comes down to 2 large enums and an object type...

**Comment by DanielRosenwasser** [comment]
So one thing that I know will help is combining your enums into one.

Otherwise, we don't currently have a good heuristic to avoid triggering the subtype reduction check. I did try the following really naive change:

```diff
         function addTypeToUnion(typeSet: Type[], includes: TypeFlags, type: Type) {
             const flags = type.flags;
-            if (flags & TypeFlags.Union) {
+            if (flags & TypeFlags.Union && !(flags & TypeFlags.EnumLike)) {
                 return addTypesToUnion(typeSet, includes, (<UnionType>type).types);
```

but for some reason I had issues running locally in parallel.

Maybe @ahejlsberg or @weswigham have some ideas here.
